File: php/tests/StringConstraintsTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\ErrorCode;
use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * §6.1, §9.3 — String constraints.
 */
class StringConstraintsTest extends TestCase
{
    use TestHelpers;

    private function errs(mixed $value, array $constraints): array
    {
        $r = new Registry();
        $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/str-test',
            'types' => [
                'Str' => ['properties' => ['v' => array_merge(['type' => 'string'], $constraints)]],
            ],
        ]);
        $schema = $r->getSchema('https://example.org/schemas/str-test');
        return validate(['_type' => 'Str', 'v' => $value], $schema->types['Str'], $schema, $r);
    }

    public function test_min_length_ok(): void
    {
        $this->assertSame([], $this->errs('hi', ['minLength' => 2]));
    }

    public function test_min_length_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs('x', ['minLength' => 2]), ErrorCode::STRING_TOO_SHORT));
    }

    public function test_max_length_ok(): void
    {
        $this->assertSame([], $this->errs('hi', ['maxLength' => 5]));
    }

    public function test_max_length_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs('toolong', ['maxLength' => 5]), ErrorCode::STRING_TOO_LONG));
    }

    public function test_pattern_match(): void
    {
        $this->assertSame([], $this->errs('abc123', ['pattern' => '^[a-z]+[0-9]+$']));
    }

    public function test_pattern_no_match(): void
    {
        $this->assertTrue($this->hasCode($this->errs('123abc', ['pattern' => '^[a-z]+[0-9]+$']), ErrorCode::PATTERN_MISMATCH));
    }

    public function test_enum_match(): void
    {
        $this->assertSame([], $this->errs('yes', ['enum' => ['yes', 'no']]));
    }

    public function test_enum_mismatch(): void
    {
        $this->assertTrue($this->hasCode($this->errs('maybe', ['enum' => ['yes', 'no']]), ErrorCode::ENUM_MISMATCH));
    }

    public function test_type_mismatch(): void
    {
        $this->assertTrue($this->hasCode($this->errs(42, []), ErrorCode::TYPE_MISMATCH));
    }
}

File: java/src/test/java/org/oojs/StringConstraintsTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class StringConstraintsTest extends TestHelpers {

    private List<ValidationError> errs(Object value, Map<Object, Object> constraints) {
        Registry r = new Registry();
        Map<Object, Object> propDef = map("type", "string");
        propDef.putAll(constraints);
        r.loadMap(map("$oojs", "1.0", "$id", "https://example.org/schemas/str-test",
                "types", map("Str", map("properties", map("v", propDef)))));
        Schema schema = r.getSchema("https://example.org/schemas/str-test");
        return new Validator(r).validate(map("_type", "Str", "v", value), schema.types.get("Str"), schema);
    }

    @Test void min_length_ok()   { assertEquals(List.of(), errs("hi", map("minLength", 2))); }
    @Test void min_length_fail() { assertTrue(hasCode(errs("x", map("minLength", 2)), ErrorCode.STRING_TOO_SHORT)); }
    @Test void max_length_ok()   { assertEquals(List.of(), errs("hi", map("maxLength", 5))); }
    @Test void max_length_fail() { assertTrue(hasCode(errs("toolong", map("maxLength", 5)), ErrorCode.STRING_TOO_LONG)); }
    @Test void pattern_match()   { assertEquals(List.of(), errs("abc123", map("pattern", "^[a-z]+[0-9]+$"))); }
    @Test void pattern_no_match(){ assertTrue(hasCode(errs("123abc", map("pattern", "^[a-z]+[0-9]+$")), ErrorCode.PATTERN_MISMATCH)); }
    @Test void enum_match()      { assertEquals(List.of(), errs("yes", map("enum", list("yes", "no")))); }
    @Test void enum_mismatch()   { assertTrue(hasCode(errs("maybe", map("enum", list("yes", "no"))), ErrorCode.ENUM_MISMATCH)); }
    @Test void type_mismatch()   { assertTrue(hasCode(errs(42, map()), ErrorCode.TYPE_MISMATCH)); }
}
